File: src/my_cpp_pkg/src/add_two_ints_client.cpp
#include "example_interfaces/srv/add_two_ints.hpp"
#include "rclcpp/rclcpp.hpp"

using example_interfaces::srv::AddTwoInts;
using std::placeholders::_1;
using std::placeholders::_2;

class AddTwoIntsClientNode : public rclcpp::Node {
public:
  AddTwoIntsClientNode() : Node("add_two_ints_client") {

    threads_.push_back(std::thread(
        std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 2, 2)));
  }

  void callAddTwoIntsService(int a, int b) {
    auto client = this->create_client<AddTwoInts>("add_two_ints");
    while (!client->wait_for_service(std::chrono::seconds(1)))
      RCLCPP_WARN(this->get_logger(), "Waiting for the server...");

    auto request = std::make_shared<AddTwoInts::Request>();
    request->a = a;
    request->b = b;

    auto future = client->async_send_request(request);
    try {
      auto response = future.get();
      RCLCPP_INFO(this->get_logger(), "%d + %d = %d", a, b, (int)response->sum);
    } catch (const std::exception &e) {
      RCLCPP_ERROR(this->get_logger(), "Error while calling service");
    }
  }

private:
  std::vector<std::thread> threads_;
};

int main(int argc, char const *argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<AddTwoIntsClientNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: src/my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3
# flake8: noqa
"""
This is a simple node that publishes a string message to the topic "my_first_topic"
"""


import rclpy
from rclpy.node import Node
from functools import partial
from example_interfaces.srv import AddTwoInts


class AddTwoIntsClientNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")
        self.call_add_two_ints_server(3, 7)

    def call_add_two_ints_server(self, a, b):
        client = self.create_client(AddTwoInts, "add_two_ints")
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for Server Add Two Ints...")

        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        future = client.call_async(request=request)
        future.add_done_callback(partial(self.callback_call_add_two_ints, a=a, b=b))

    def callback_call_add_two_ints(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(f"{a} + {b} = {response.sum}")
        except Exception as e:
            self.get_logger().error(f"Service call failed with unexpected error: {e}")


def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClientNode()
    rclpy.spin(node)  # mantém o programa rodando
    rclpy.shutdown()


if __name__ == "__main__":
    main()
